### PR TITLE
Global setLogger was added in Dispatcher, if Logger was configured in…

### PR DESCRIPTION
**Pull request by lexizz**
…side Dispatcher #18840

| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Tests pass?   | ✔️
| Fixed issues  | #18840

Test `tests/framework/log/DispatcherTest.php` method `testConfigureLogger` updated

### CODE REVIEW

**Comment by bizley** [comment]
The problem here is that we allow to configure logger from the dispatcher when it obviously should be configured at the top level of application. But it's too late for that change.

File: framework/log/Dispatcher.php [hunk 1]
```diff
@@ -123,12 +124,15 @@ public function getLogger()
      * Sets the connected logger.
      * @param Logger|string|array $value the logger to be used. This can either be a logger instance
      * or a configuration that will be used to create one using [[Yii::createObject()]].
+     * @throws InvalidConfigException
      */
     public function setLogger($value)
     {
         if (is_string($value) || is_array($value)) {
             $value = Yii::createObject($value);
+            Yii::setLogger($value);
```
Comment (bizley) [inline framework/log/Dispatcher.php:133]: I think this should be moved below so already configured object can also be set at the main logger.

Comment (samdark) [reply framework/log/Dispatcher.php:133]: Agree. Right after `$this->_logger = $value;`

Comment (lexizz) [reply framework/log/Dispatcher.php]: Ok. Thanks

**Review by samdark — APPROVED** [review]

File: framework/CHANGELOG.md [hunk 2]
```diff
@@ -8,6 +8,7 @@ Yii Framework 2 Change Log
 - Enh #18826: Add ability to turn the sorting off for a clicked column in GridView with multisort (ditibal)
 - Bug #18646: Remove stale identity data from session if `IdentityInterface::findIdentity()` returns `null` (mikehaertl)
 - Bug #18832: Fix `Inflector::camel2words()` adding extra spaces (brandonkelly)
+- Bug #18840: Fix `yii2/log/Logger` is impossible to configure, if `YII_DEBUG = true` (lexizz)
```
Comment (bizley) [inline framework/CHANGELOG.md:11]:
```suggestion
- Bug #18840: Fix `yii2/log/Logger` to set logger globally (lexizz)
```

Comment (lexizz) [reply framework/CHANGELOG.md]: fix

**Review by bizley — APPROVED** [review]

**Review by Arhell — APPROVED** [review]

**Comment by samdark** [comment]
Merged. Than you!

**Comment by rob006** [comment]
This change was already made in https://github.com/yiisoft/yii2/pull/17477, and then reverted, because it breaks BC and escalates magic behavior. Please revert it (again :P).

**Comment by bizley** [comment]
Ouu, you always come after the fact :D Thanks rob006 ;) I'll revert. Sorry @lexizz you have to set logger on the top level of application.

**Comment by rob006** [comment]
> Ouu, you always come after the fact :D

Yeah, sorry about that, but I don't have time to watch this repo and I track only new commits, so my reviews are always too late. :/

**Comment by bizley** [comment]
I'm kidding 😀 thanks for keeping your finger on the pulse

**Comment by lexizz** [comment]
1. Is it"BC"?
2. "you have to set logger on the top level of application" - Do you mean in DI container?

**Comment by rob006** [comment]
@lexizz It means before app is created and run. Something like:

```
Yii::setLogger($loggerConfig);
(new yii\web\Application($config))->run();
```

BC stands for backward compatibility. This change significantly changes behavior in some scenarios, you don't want this in framework, because people will avoid updating it if every upgrade break something.

**Comment by lexizz** [comment]
@rob006 
```
Yii::setLogger($loggerConfig);
(new yii\web\Application($config))->run();
```
This is not an obvious behavior. 
Configuring Logger before initializing the application is not very good.

**Comment by lexizz** [comment]
@rob006 How does my edit break backward compatibility?

**Comment by bizley** [comment]
@lexizz you can set it up in $config with `logger` key.

**Comment by lexizz** [comment]
> @lexizz you can set it up in $config with `logger` key.

How does my edit break backward compatibility?

**Comment by lexizz** [comment]
> There might be cases when developers don't want their global logger config to be changed by configuring dispatcher's logger.

If they do not configure the logger, then nothing will break, the global logger will be used.

**Comment by lexizz** [comment]
> @lexizz you can set it up in $config with `logger` key.

image

If that's what you mean, then your offer won't work.

https://github.com/yiisoft/yii2/issues/18840

**Comment by bizley** [comment]
Hmm, yes, I've just realized there is no option to create object from array in Yii::setLogger() which would be nice now to have... Should we add this feature? I can prepare it.

**Comment by rob006** [comment]
>  Configuring Logger before initializing the application is not very good.

Well, you don't have much of a choice - you need logger before app is instantiated (so you could log events before app and dispatcher is ready), if you use config passed to `Application` constructor, then it is already too late to configure logger.

**Comment by bizley** [comment]
Is it too late? Need to check this.

**Comment by lexizz** [comment]
@bizley 
> Hmm, yes, I've just realized there is no option to create object from array in Yii::setLogger() which would be nice now to have... Should we add this feature? I can prepare it.

Yes. It would be very good

**Comment by bizley** [comment]
If there are no objections to add the ability to set logger with array I'll add PR tomorrow morning.

**Comment by rob006** [comment]
> Hmm, yes, I've just realized there is no option to create object from array in Yii::setLogger() which would be nice now to have... Should we add this feature? I can prepare it.

I'm not sure if we need this. Yii uses config arrays mostly due to config merging or delaying instantiating object, but this is not the case, it should be more clean to use `new Logger($config))` instead of array with config.

**Comment by lexizz** [comment]
> Well, you don't have much of a choice - you need logger before app is instantiated (so you could log events before app and dispatcher is ready), if you use config passed to `Application` constructor, then it is already too late to configure logger.

Why configure Logger before initializing the application, when it is possible in the DI container. But this is very inconvenient and not obvious))

**Comment by lexizz** [comment]
@rob006 You didn't answer the question))
> How does my edit break backward compatibility?

**Comment by rob006** [comment]
@lexizz Example is in PR I linked in my first comment: https://github.com/yiisoft/yii2/pull/17477#issuecomment-516501680.

**Comment by lexizz** [comment]
@rob006 
> @lexizz Example is in PR I linked in my first comment: [#17477 (comment)](https://github.com/yiisoft/yii2/pull/17477#issuecomment-516501680).

As far as I understand, you said in  [#17477 (comment)] that the global logger should not be overridden if the local one is configured. But we usually expect that when setting up a logger, the global one will behave the same as the local one, right? Therefore, I do not agree with your statement, which is in  [#17477 (comment)]))

When is the right time to make such changes?)) Never?

**Comment by bizley** [comment]
We already provided you with the proper solution for your case.

**Comment by rob006** [comment]
> But we usually expect that when setting up a logger, the global one will behave the same as the local one, right? 

No, I don't expect that creating local object would change logger used by `Yii::getLogger()`. It is counter-intuitive in terms of OOP.

> When is the right time to make such changes?)) Never?

Exactly never, because it is a step in a wrong direction - it is unexpected that creating local object would affect completely unrelated object or some global state. Better solution would be to create some kind of `Yii::initializeApp($config)` helper, that would use app config to configure logger used by `Yii` and then create app object. It could also handle configuring other static variables and state - it is already controversial that instantiating `Application` would magically overwrite `Yii::$app` and setup some global state, but it can't be changed without major release (which will probably never happen).

**Comment by lexizz** [comment]
@bizley @rob006 Ok. Thanks. I realized.

**Pull request merged**